### PR TITLE
Show global errors

### LINKED ISSUE #153: Show a demo indicator
We should somehow indicate that the user is using this app in demo mode, e.g. via a Banner, MaterialBanner, or above/below the footer

### LINKED ISSUE #172: Create framework for displaying persistent banners at the top
**Is your feature request related to a problem? Please describe.**
There are certain conditions where the user should be notified about some persistent condition, for example, if you're logged in with a demo account, you have no internet connection or your token is about to expire or is expired.

**Describe the solution you'd like**
Persistent banners at the top of the screen would be cool.

### LINKED ISSUE #259: Display global errors

### PR DESCRIPTION
**Pull request by MarcelGarus**
**Closes: #259, closes: #172, closes: #153**

Now, global errors are shown at the top.
For now, global errors include having no internet access and having an expired authentication token.
Also, demo accounts are now indicated by a demo tag instead of an account avatar at the top right.

**Screenshots**

image

### CODE REVIEW

File: lib/app/services/banner.dart [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import 'package:flutter/foundation.dart';
+import 'package:get_it/get_it.dart';
+
+class Banner {
```
Comment (JonasWanke) [inline lib/app/services/banner.dart:5]:
```suggestion
@immutable
class Banner {
```

File: lib/app/services/banner.dart [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  const Banner(this.name);
```
Comment (JonasWanke) [inline lib/app/services/banner.dart]:
```suggestion
  const Banner(this.name) : assert(name != null);
```

File: lib/app/widgets/banners.dart [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import 'package:banners/banners.dart';
+import 'package:flutter/material.dart';
+import 'package:schulcloud/app/app.dart';
+
+class OfflineBanner extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return BannerScaffold(
+      backgroundColor: Theme.of(context).errorColor.withOpacity(0.5),
```
Comment (JonasWanke) [inline lib/app/widgets/banners.dart]:
```suggestion
      backgroundColor: context.theme.errorColor.withOpacity(0.5),
```

File: lib/app/widgets/account_avatar.dart [hunk 4]
```diff
@@ -1,49 +1,73 @@
...
         final backgroundColor =
-            user?.avatarBackgroundColor ?? context.theme.primaryColor;
-
+            snapshot.data?.avatarBackgroundColor ?? context.theme.primaryColor;
         return CircleAvatar(
           backgroundColor: backgroundColor,
           maxRadius: 16,
           child: Text(
-            user?.avatarInitials ?? '…',
+            snapshot.data?.avatarInitials ?? snapshot.hasError ? 'X' : '…',
```
Comment (JonasWanke) [inline lib/app/widgets/account_avatar.dart]: `??` has a higher precedence than `e1 ? e2 : e3`: https://dart.dev/guides/language/specifications/DartLangSpec-v2.2.pdf#page=199
```suggestion
            snapshot.data?.avatarInitials ?? (snapshot.hasError ? 'X' : '…'),
```

File: lib/main.dart [hunk 5]
```diff
@@ -102,22 +109,28 @@ Future<void> main({AppConfig appConfig = schulCloudAppConfig}) async {
...
+
+    // Set demo banner based on current user.
+    // TODO(marcelgarus): dipose id stream.
+    services.storage.userIdString
+        .map((idString) => Id<User>(idString))
+        .listen((userId) {
+      userId.resolve().listen((user) {
+        // TODO(marcelgarus): Don't hardcode role id.
+        final isDemo =
+            user?.roleIds?.contains(Id<Role>('0000d186816abba584714d02')) ??
```
Comment (JonasWanke) [inline lib/main.dart]: This is the role ID of a demo student. We also need that of a demo teacher: 0000d186816abba584714d03
Source: https://github.com/schul-cloud/schulcloud-server/blob/a2e15cb617916589cabbe18c0deea8499ed61080/backup/setup/roles.json#L301-L342

Or (even better): Check transitively included roles for `demo` (0000d186816abba584714d00): https://github.com/schul-cloud/schulcloud-server/blob/a2e15cb617916589cabbe18c0deea8499ed61080/backup/setup/roles.json#L248-L278

File: lib/l10n/intl_en.arb [hunk 6]
```diff
@@ -1,5 +1,6 @@
 {
   "@@locale": "en",
+  "app_demo": "demo",
```
Comment (JonasWanke) [inline lib/l10n/intl_en.arb]:
```suggestion
  "app_demo": "Demo",
```

File: lib/app/widgets/banners.dart [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+      body: Padding(
+        padding: EdgeInsets.all(8),
```
Comment (JonasWanke) [inline lib/app/widgets/banners.dart]: I'd prefer a horizontal padding of 16 so it matches our globally used horizontal padding (i.e., that of the AppBar title and list items). Maybe decrease the vertical padding to 4?

(Also applies to the `TokenExpiredBanner` below.)

File: lib/app/widgets/banners.dart [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        child: Text(context.s.app_error_noConnection),
+      ),
+    );
+  }
+}
+
+class TokenExpiredBanner extends StatelessWidget {
```
Comment (JonasWanke) [inline lib/app/widgets/banners.dart:22]: We might want to include a sign-out-icon here directly so the user doesn't have to search for it

**Review by JonasWanke — CHANGES_REQUESTED** [review]
Maybe we should add a more descriptive demo-banner in the `AccountDialog` so after tapping "demo" the user also gets a short explanation? (e.g., the banner text of the web-client)

**Comment by MarcelGarus** [comment]
There's now a sign out button in the `TokenExpiredBanner` and a demo explanation in the `AccountDialog`:

sign out button | demo explanation
---|---
image | image

File: lib/l10n/intl_en.arb [hunk 9]
```diff
@@ -1,6 +1,7 @@
 {
   "@@locale": "en",
-  "app_demo": "demo",
+  "app_demo": "Demo",
+  "app_demo_explanation": "This is a demo account. You may not be able to perform certain actions, like uploading files.",
```
Comment (JonasWanke) [inline lib/l10n/intl_en.arb]: Fun fact: uploading files is the one write-operation you _can_ perform as a demo user: https://github.com/schul-cloud/schulcloud-server/blob/a2e15cb617916589cabbe18c0deea8499ed61080/backup/setup/roles.json#L267

Maybe keep it general like "This is a demo account. All actions that create or change data are deactivated and not visible." (translated from the web-client) and ignore that exception?

(This also applies to the German translation.)

File: lib/app/widgets/banners.dart [hunk 10]
```diff
@@ -21,8 +25,21 @@ class TokenExpiredBanner extends StatelessWidget {
...
+        padding: _bannerPadding,
+        child: Row(
+          children: <Widget>[
+            Expanded(child: Text(context.s.app_error_tokenExpired)),
+            SizedBox(width: 16),
+            IconButton(
+              onPressed: () => signOut(context),
+              icon: SvgPicture.asset(
+                'assets/icon_signOut.svg',
+                color: context.theme.isDark ? Colors.white : Colors.black,
```
Comment (JonasWanke) [inline lib/app/widgets/banners.dart]: black_hole_flutter to the rescue!
```suggestion
                color: context.theme.contrastColor,
```

**Review by JonasWanke — CHANGES_REQUESTED** [review]

**Review by JonasWanke — APPROVED** [review]

**Pull request merged**